src/renderer/wg_engine/tvgWgGeometry.h

```diff
@@ -24,102 +24,435 @@
-    WgPoint operator / (const float c) const { return { x / c, y / c }; }
-
-    WgPoint negative() const { return {-x, -y}; }
-    inline void negate() { x = -x; y = -y; }
-    inline float length() const { return sqrt(x*x + y*y); }
-    inline float length2() const { return x*x + y*y; }
-    inline float dot(const WgPoint& p) const { return x * p.x + y * p.y; }
-    inline float dist(const WgPoint& p) const { return sqrt(dist2(p)); }
-    inline float dist2(const WgPoint& p) const { return ((p.x - x)*(p.x - x) + (p.y - y)*(p.y - y)); }
-    inline bool equal(const WgPoint& p) const { return tvg::equal(x, p.x) && tvg::equal(y, p.y); }
-    inline void normalize() { float rlen = 1.0f / length(); x *= rlen; y *= rlen; }
-    inline WgPoint normal() const { float rlen = 1.0f / length(); return { x * rlen, y * rlen }; }
-    inline WgPoint lerp(const WgPoint& p, float t) const { return { x + (p.x - x) * t, y + (p.y - y) * t }; };
-    inline WgPoint trans(const Matrix& m) const { return { x * m.e11 + y * m.e12 + m.e13, x * m.e21 + y * m.e22 + m.e23 }; };
-};
-
-struct WgMath
-{
-    Array<float> sinus;
-    Array<float> cosin;
-    bool initialized{};
 
-    void initialize();
-    void release();
-};
+// base vector operations
+static inline Point operator-(const Point& a) { return {-a.x, -a.y}; }
+static inline float length2(const Point& a) { return a.x*a.x+a.y*a.y; };
+static inline float distance2(const Point& a, const Point& b) { return length2(a - b); };
+static inline float distance(const Point& a, const Point& b) { return length(a - b); };
+static inline float dot(const Point& a, const Point& b) { return a.x*b.x + a.y*b.y; };
+static inline Point min(const Point& a, const Point& b) { return { std::min(a.x, b.x), std::min(a.y, b.y) }; };
+static inline Point max(const Point& a, const Point& b) { return { std::max(a.x, b.x), std::max(a.y, b.y) }; };
+static inline Point lerp(const Point& a, const Point& b, float t) { return a * (1.0f - t) + b * t; };
+static inline Point normalize(const Point& a) { float rlen = 1.0f / length(a); return { a.x * rlen, a.y * rlen }; }
+static inline Point hat(const Point& a) { return { +a.y, -a.x }; }
 
+// default size of vertex and index buffers
+#define WG_POINTS_COUNT 16384
 
-struct WgPolyline
-{
-    Array<WgPoint> pts;
-    Array<float> dist;
-    // polyline bbox points indexes
-    uint32_t iminx{};
-    uint32_t iminy{};
-    uint32_t imaxx{};
-    uint32_t imaxy{};
-    // total polyline length
-    float len{};
+// simple vertex buffer
+struct WgVertexBuffer {
+    Point vbuff[WG_POINTS_COUNT]; // vertex buffer
+    float vdist[WG_POINTS_COUNT]; // distance to previous point
+    float vleng[WG_POINTS_COUNT]; // distance to the first point through all previous points
+    size_t vcount{};
     bool closed{};
 
-    WgPolyline();
+    // callback for external process of polyline
+    using onPolylineFn = std::function<void(const WgVertexBuffer& buff)>;
+
+    // reset buffer
+    inline void reset() {
```

Comment:
please remove `inline` in those class methods. Both syntaxes for inlining functions (using explicit inline and defining member-function inside class definition) provides only hint about inlining for compiler. From performance point of view, they are equal.